docs: add KeptnTask ref page; enhance guide chapter for keptn-no-k8s (#2103)
keptn/lifecycle-toolkit: docs/content/en/docs/yaml-crd-ref/task.md

---
title: KeptnTask
description: Define a run of a KeptnTaskDefinition
weight: 85
---

When using Keptn to run tasks for software
that is deployed outside of Kubernetes,
you must create the `KeptnTask` resource manually
and modify it manually for each new run.
Keptn automatically populates the `KeptnTask` resource

> for tasks that deploy software on Kubernetes.

Author: done

## Synopsis

```yaml
apiVersion: lifecycle.keptn.sh/v1alpha3
kind: KeptnTask
metadata:
  name: <name-of-this-run>
spec:
  taskDefinition: <name-of-KeptnTaskDefinition resource>
  context:
    appName: "<name-of-KeptnApp-resource>"
    appVersion: "1.0.0"
    objectType: ""
    taskType: ""
    workloadName: "my-workload"
    workloadVersion: "1.0.0"
  parameters: <parameters to pass to job>
  secureParameters: <secure parameters to pass to job>
  checkType: ""
  retries: <integer>
  timeout: <duration-in-seconds>
```

## Fields

* **apiVersion** -- API version being used.

> `

Author: done

* **kind** -- Resource type.

>    Must be set to `KeptnTask`

Author: done

---
title: KeptnTask
description: Define a run of a KeptnTaskDefinition
weight: 85
---

When using Keptn to run tasks for software
that is deployed outside of Kubernetes,

> you must create the `KeptnTask` definition manually

Reviewer: ```suggestion
you must create the `KeptnTask` resource manually
```

Keptn automatically populates the `KeptnTask` resource
for tasks that deploy software on Kubernetes.

## Synopsis

```yaml
apiVersion: lifecycle.keptn.sh/v1alpha3
kind: KeptnTask
metadata:
  name: <name-of-this-run>
spec:
  taskDefinition: <name-of-KeptnTaskDefinition resource>
  context:
    appName: "<name-of-KeptnApp-resource>"
    appVersion: "1.0.0"
    objectType: ""
    taskType: ""
    workloadName: "my-workload"
    workloadVersion: "1.0.0"
  parameters: <parameters to pass to job>
  secureParameters: <secure parameters to pass to job>
  checkType: ""
  retries: <integer>
  timeout: <duration-in-seconds>
```

## Fields

* **apiVersion** -- API version being used.
`
* **kind** -- Resource type.
   Must be set to `KeptnTask`

* **metadata**
  * **name** -- Unique name of this run of the task.
    This name must be modified each time you run this `KeptnTask`,
    so a common practice is to add a number to the end of the string
    so you can increment the number for each run.
    Names must comply with the
    [Kubernetes Object Names and IDs](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names)
    specification.

> * **spec** - Defines the desired state of this `KeptnTask` resource

Reviewer: ```suggestion
* **spec** - Defines the speficication of this `KeptnTask` resource
```

>   * **workload** - Name of the
>       [KeptnWorkload](../crd-ref/lifecycle/v1alpha3/#keptnworkload)
>       resource for which the `KeptnTask` is being executed.
> 
>   * **workloadVersion** - `KeptnWorkload` version
>       for which the `KeptnTask` is being executed.
>   * **appName** - Name of the
>       [KeptnApp](../yaml-crd-ref/app.md) resource
>       for which the `KeptnTask` is being executed.
>   * **appVersion** - Version of the `KeptnApp` resource
>       for which the `KeptnTask` is being executed.
> 

Reviewer: We do not support anymore these parameters in v1alpha3, see here https://github.com/keptn/lifecycle-toolkit/blob/main/lifecycle-operator/apis/lifecycle/v1alpha3/keptntask_types.go#L34

```suggestion
```

  * **taskDefinition** - Name of the corresponding `KeptnTaskDefinition` resource.
    This `KeptnTaskDefinition` can be located in the same namespace

>     or in the `Keptn` name space.

Reviewer: ```suggestion
    or in the Keptn installation namespace.
```

  * **context** - Contextual information about the task execution
    * **appName** - Name of the
      [KeptnApp](../yaml-crd-ref/app.md) resource
      for which the `KeptnTask` is being executed.
    * **appVersion** - Version of the `KeptnApp` resource
      for which the `KeptnTask` is being executed.

    * **objectType** - Indicates whether this `KeptnTask`
      is being executed for a `KeptnApp` or a `KeptnWorkload` resource.
      When populating this resource manually
      to run a task for a non-Kubernetes deployment,
      set this value to `""`:
      Keptn populates this field based on annotations
      to the `KeptnWorkload` and `KeptnApp` resources.

    * **taskType** Indicates whether this `KeptnTask`
      is part of the pre- or post-deployment phase.
      When populating this resource manually

>       to run a task for a non-Kubernetes deployment,,
> u     set this value to `""`:

Reviewer: ```suggestion
      to run a task for a non-Kubernetes deployment,
      set this value to `""`:
```

      Keptn populates this field based on annotations
      to the `KeptnWorkload` and `KeptnApp` resources.

>     * **workloadName** - Name of the `KeptnWorkload`
>       for which the `KeptnTask` is being executed
>     * **workloadVersion** - Version of the `KeptnWorkload`
>       for which the `KeptnTask` is being executed

Reviewer: ```suggestion
    * **workloadName** - Name of the `KeptnWorkload`
      for which the `KeptnTask` is being executed.
    * **workloadVersion** - Version of the `KeptnWorkload`
      for which the `KeptnTask` is being executed.
```

  * **parameters** (optional) -- Parameters that are passed to the job
    that executes the `KeptnTask`.
  * **secureParameters** (optional) -- Secure parameters that are passed
    to the job that executes the `KeptnTask`.

>     These are stored and accessed as secrets in the cluster.

Reviewer: ```suggestion
    These are stored and accessed as Kubernetes `Secrets` in the cluster.
```

---
title: KeptnTask
description: Define a run of a KeptnTaskDefinition
weight: 85
---

Keptn populates the `KeptnTask` resource
for tasks that deploy software on Kubernetes.

> When using Keptn to deploy software built on a virtual machine,

Reviewer 2: Perhaps just say "When using Keptn to run tasks for software that is deployed outside of kubernetes". This method would work for things deployed on anything from mainframes to VMs to serverless etc.

Author: Yes, this is much better -- done.

you must manually populate the `KeptnTask` definition
and modify it manually for each new run.

## Synopsis

```yaml
apiVersion: lifecycle.keptn.sh/v1alpha3
kind: KeptnTask
metadata:
  name: <name-of-this-run>
spec:
  workload: "my-workload"
  workloadVersion: "1.0.0"

>   appVersion: "1.0.0"

Reviewer 2: As mentioned above remove these fields

      TODO: API reference shows `appName` and `appVersion` in `TaskContext`

      TODO: API reference shows `parameters` and `secureParameters` here.
      Should they be added to this list?

      TODO: API reference also shows `CheckType`, `retries`, and `timeout` here.

  * **taskDefinition** - Name of the corresponding `KeptnTaskDefinition` resource.
  This `KeptnTaskDefinition` can be located in the same namespace
  or in the Keptn name space.
    * **context** - Contextual information about the task execution
      * **appName** - Name of the
          [KeptnApp](../yaml-crd-ref/app.md) resource
          for which the `KeptnTask` is being executed.
      * **appVersion** - Version of the `KeptnApp` resource
          for which the `KeptnTask` is being executed.

          TODO: So I can create a `KeptnApp` resource that lists multiple workloads
          to be processed?
           Do I need to annotate the Workloads
          and can I autocreate the `KeptnApp` or do I need to do it manually?
      * **objectType** - Indicates whether this `KeptnTask`
          is being executed for a `KeptnApp` or a `KeptnWorkload` resource.

>           TODO: So why is this set to null (`""`) in the example?

Reviewer 2: Because I can't find where the suitable types for this field are documented and an empty string just happens to work. This is one for whomever coded that piece.

Author: Okay.  I'm removing the TODO's from this page so we can get it merged.  We can improve it later.

      * **taskType** Indicates whether this `KeptnTask`
          is part of the pre- or post-deployment phase

>           TODO: So why is this set to null (`""`) in the example?

Reviewer 2: See above

      * **workloadName** - Name of the `KeptnWorkload`
          for which the `KeptnTask` is being executed
      * **workloadVersion** - Version of the `KeptnWorkload`
          for which the `KeptnTask` is being executed

>           TODO: Why are these fields both here and directly in `spec` above?

Reviewer 2: The spec was wrong and has now been fixed. We can remove all `spec.*` things. Everything now lives in `spec.context.*`

Author: Docs are now reconciled with revised v1alpha3 spec

## Usage

Applying this file causes Keptn to create a Job and a Pod
and run the associated `KeptnTaskDefinition`.

Use the following commands to show the current status of the jobs:

```shell
kubectl get keptntasks
kubectl get pods
```

Each time you want to rerun the `KeptnTask` resource,
you must update the following fields
to create a unique `KeptnTask` resource:

* `name`
* `spec.appVersion`

> * `spec.workloadVersion`

Reviewer 2: Remove

* `spec.context.appVersion`
* `spec.context.workloadVersion`

A common practice is to just increment the value of each field.

## Examples

For this example, the `KeptnTaskDefinition` resource is:

```yaml
apiVersion: lifecycle.keptn.sh/v1alpha3
kind: KeptnTaskDefinition 
metadata:
  name: helloworldtask
spec:
  retries: 0
  timeout: 30s
  container:
    name: cowsay
    image: rancher/cowsay:latest
    args:
      - 'hello world'
```

And the `KeptnTask` resource for the first run is:

```yaml
apiVersion: lifecycle.keptn.sh/v1alpha3
kind: KeptnTask
metadata:
  name: runhelloworld1
spec:
  workload: "my-workload"
  workloadVersion: "1.0.0"
  appVersion: "1.0.0"

>   app: "my-app"

Reviewer 2: Remove

Author: done

  taskDefinition: helloworldtask
  context:
    appName: "my-app"
    appVersion: "1.0.0"
    objectType: ""
    taskType: ""
    workloadName: "my-workload"
    workloadVersion: "1.0.0"
```

The `KeptnTask resource for the second run is:

```yaml
apiVersion: lifecycle.keptn.sh/v1alpha3
kind: KeptnTask
metadata:
  name: runhelloworld2
spec:
  workload: "my-workload"
  workloadVersion: "1.0.1"
  appVersion: "1.0.1"

>   app: "my-app"

Reviewer 2: Remove

Author: done
